### ISSUE TITLE
Minimum GCC 4.9.0

### ISSUE THREAD

**Issue by endJunction** [ufz/ogs#1084]
Some major changes to the STL (regarding the c++11 standard) were implemented in the GCC version 4.9.0 and are not included in the currently required version 4.8.1.

The changes include updates of the STL containers' interfaces, _e.g._ the `const_iterator` for the `erase` function, but many more others. Unfortunately not reflected in their changelog, but see for example this [commit message](https://github.com/gcc-mirror/gcc/commit/6b0e18ca48bb4b4c01e7b5be2b98849943fdcf91) in github/gcc-mirror.

The changes do reflect the requirements of the c++11 standard; for example the `std::list::erase(iterator)` is **replaced** by `std::list::erase(const_iterator)`.

TODO:
- [x] Update travis.
- [ ] Update modules on envinf
- [ ] VTK (vtk-7.1.0 does not work with gcc-6.2.0/5.4.0 on envinf; linking errors referring to glibc-3.4.15.)
- [ ] Maybe PETSc, but I didn't check.

Not sure if all this recompilations are necessary on conan, but if yes this is the initial list:
- [ ] VTK
- [ ] Qt
- [ ] PETSc
- [ ] LIS
- [ ] Boost
- [ ] ...

**Comment by norihiro-w** [comment]
Could you give us more details? What changes in the new version do we need to use?

**Comment by chleh** [comment]
I'm skeptical, too. Don't we have more serious problems than the perfect const correctness in #1075?

**Comment by TomFischer** [comment]
@chleh You are right, there are many places under construction in OGS-6 development. But why should we not use newer library versions to work on these open tasks that make the code better and programming more comfortable?

Is it the time effort that you afraid of or do you have other concerns? Arch linux can install a very new gcc version. I guess ubuntu users with installed new LTS ubuntu version (2016-04) should be able to install a newer gcc. On envinf1 we can load an even newer gcc module. Maybe only on envinf Qt, VTK and PETSc have to be recompiled once and maybe the WKDV can do this for us?

We have a similar situation with the DataExplorer. The DE version is based on an old version of Qt (4.8) while version 5.5.1 was released Oct 2015. When we decide to set the minimum gcc version to 4.9 (**or newer**) we should also discuss to switch to a newer Qt version. So the update work has to be done only once.

**Comment by chleh** [comment]
It's simply the time effort. I agree, that @endJunction's intentions are good, but I'd give this particular issue a rather low priority. But that's only my personal opinion.

Regarding Qt5: That makes things worse. @endJunction already spent much time on porting the data explorer, and AFAIK he got stuck. I know that support for Qt4 is going to end, that's bad. I don't know what to do best. Wait until it hurts? Wait until somebody really wants to migrate? Maybe we shouldn't discuss that here. At least, if somebody starts the migration, there should be a time limit and if the one gets stuck, the migration should be stopped or the problems be analyzed.

**Comment by endJunction** [comment]
@norihiro-w I updated the issue text adding some explanations.

**Comment by norihiro-w** [comment]
thanks!

**Comment by chleh** [comment]
Still thinking we don't need the new lib. :smile:

**Comment by endJunction** [comment]
The minimum gcc version for language is 4.8, but for the STL it is version 5.0; [changelog](https://gcc.gnu.org/gcc-5/changes.html). The 4.9 changelog does not explicitly mention full or complete c++11 support but improvements.

**Comment by endJunction** [comment]
Just another idea for the linux we could stick with the newest LTS version of ubuntu which is 16.04 and will be released on 21. April.
It would be then [gcc-5.3](http://packages.ubuntu.com/xenial/devel/gcc) and [clang-3.8](http://packages.ubuntu.com/xenial/devel/clang)

**Comment by TomFischer** [comment]
:+1: for ubuntu 16.04 compiler versions.

**Comment by chleh** [comment]
As already said many times: I think we should change to new versions of compiler/libs only if we really urgently have to, e.g., in the case of serious bugs. In the future there will be users who will not want install the newest Ubuntu LTS as soon as it is released.
A usual PhD study takes at least three years, and users will probably not want to change their systems during that period. And if you are not maintaining the computers you work with, it will be very difficult to get our stuff running if we require too new stuff.
With that in mind I'd suggest at least we should support the Ubuntu LTS from two years ago, or Debian oldstable.
I'd be rather conservative in that regard and make sure that our code also works with relatively old libs and compilers. And currently g++ 4.8 obviously has all the features we need (except of the one in #1075, but I don't count that).

**Comment by endJunction** [comment]
Christoph and I have observed recently that optimized gcc-4.8.1 executables do not work correctly (on envinf) but in Debug build type. Also of all compilers this is the oldest one (being released on 2013-05-31).

One more time the suggestion is to set the minimum to ubuntu's 16.04 gcc-5.3 (and clang-3.8).

@bilke Here is something about travis: https://jonasw.de/blog/2015/07/22/develop/cplusplus14-on-travis-with-cmake/

**Comment by endJunction** [comment]
On envinf I have successfully compiled the code with gcc-6.2.0 and recompiled vtk-7.1 (OGS_LIB_VTK=Local). All other libraries were taken as is from the `cli.sh`.

**Comment by chleh** [comment]
Regarding the not correctly working executables: A single unit test fails, and the tested feature is currently probably not used in production code. I don't remember exactly whether the ctests run. Anyway, that reduces the confidence in the gcc-4.8.1 build.

In my opinion we can switch to gcc-5.3. Regarding future updates: I think we shouldn't update our minimum requirements as soon as there will be a new ubuntu LTS release, since the next LTS will already be next april.

@endJunction, it's nice that gcc-6.2.0 could compile ogs on envinf1, but in the future we should stick to testing the minimum requirements, IMHO.

**Comment by ogsbot** [comment]
OpenGeoSys development has been moved to [GitLab](https://gitlab.opengeosys.org/ogs/ogs).

[See this issue on GitLab](https://gitlab.opengeosys.org/ogs/ogs/-/issues/1084).
